numpy/core/src/umath/ufunc_object.c

```diff
@@ -817,235 +838,233 @@ get_ufunc_arguments(PyUFuncObject *ufunc,
         PyObject *key, *value;
         Py_ssize_t pos = 0;
         while (PyDict_Next(kwds, &pos, &key, &value)) {
-            Py_ssize_t length = 0;
-            char *str = NULL;
+            PyArray_Descr *dtype;
             int bad_arg = 1;
-
-#if defined(NPY_PY3K)
-            Py_XDECREF(str_key_obj);
-            str_key_obj = PyUnicode_AsASCIIString(key);
-            if (str_key_obj != NULL) {
-                key = str_key_obj;
+            int kw_id;
+            /*
+             * For maximum speed, we follow CPython's ceval.c
+             * _PyEval_EvalCodeWithName, which tries direct pointer
+             * comparison first, noting:
+             * """Speed hack: do raw pointer compares. As names are
+             *    normally interned this should almost always hit."""
+             * And we do indeed intern them.
+             */
+            for (kw_id = 0; kw_id < ufunc_badkey; kw_id++) {
+                if (ufunc_kwnames[kw_id] == key) {
+                    goto kw_found;
+                }
             }
-#endif
-
-            if (PyBytes_AsStringAndSize(key, &str, &length) < 0) {
-                PyErr_Clear();
-                PyErr_SetString(PyExc_TypeError, "invalid keyword argument");
-                goto fail;
+            /* Slow fallback, just in case */
+            for (kw_id = 0; kw_id < ufunc_badkey; kw_id++) {
+                int cmp = PyObject_RichCompareBool(key, ufunc_kwnames[kw_id], Py_EQ);
+                if (cmp > 0) {
+                    goto kw_found;
```

Comment:
Instead of using a `goto`, just make the above a helper function returning `kw_id`

Reply:
I followed CPython's `ceval` quite exactly! But with a helper function, this may become useful for `GenericReduction` as well.

Reply:
Can you link to the code you're imitating?

Reply:
https://github.com/python/cpython/blob/master/Python/ceval.c#L3759

Reply:
p.s. Do think the helper function is a good idea, but was waiting with making changes as I hoped someone would have a better idea of how to handle those name arrays and the `enum` -- I like the names but not that multiple lists have to be maintained separately.  Can one iterate over an `enum`?